### DISCUSSION TITLE
Can Only Run a Program Once, the I get an Exception

### DISCUSSION THREAD

**Discussion by pmi123** [micropython/micropython#16291] [ESP32]
I am using a ESP-WROOM-32 module. I erased the firmware, flashed it with ESP32_GENERIC-SPIRAM-20230426-v1.20.0.bin, and it works, sort of. I can access the ESP32 over /dev/ttyUSB0, load a micropython program, and run it. But only once. If I try to run it again, I get an error:
```
    MPY: soft reboot
    Traceback (most recent call last):
      File "<stdin>", line 7, in <module>
    OSError: (-259, 'ESP_ERR_INVALID_STATE')
    >>> 
```

I have to remove power from the board in order to run the same code again.

This is my code - I am cycling through the DAC values and reading them with the ADC:
```
from machine import Pin, ADC, DAC
import time

# Configure the DAC pin
dac_pin = 25
dac = DAC(Pin(dac_pin))

# Configure the ADC pin (use the correct pin number for your setup)
adc_pin = 36  # Example pin (GPIO36, also known as VP)
adc = ADC(Pin(adc_pin))

# Set the ADC attenuation (defines voltage range)
# ADC.ATTN_0DB: 0-1.1V (default)
# ADC.ATTN_2_5DB: 0-1.5V
# ADC.ATTN_6DB: 0-2.0V
# ADC.ATTN_11DB: 0-3.3V
adc.atten(ADC.ATTN_11DB)  # Set to 3.3V range

# Set ADC width (defines resolution)
# ADC.WIDTH_9BIT: 9-bit resolution (0-511)
# ADC.WIDTH_10BIT: 10-bit resolution (0-1023)
# ADC.WIDTH_11BIT: 11-bit resolution (0-2047)
# ADC.WIDTH_12BIT: 12-bit resolution (0-4095, default)
adc.width(ADC.WIDTH_12BIT)

# Output voltage from DAC and read with ADC
dac_voltage = 0
while True:
  dac_voltage += 0.25   # increase DAC voltage by 0.25V for each iteration
  dac_value = int((dac_voltage / 3.3) * 255) # convert voltage to 8 bit DAC value
  if dac_value > 256:
    break
  dac.write(dac_value)  # Write the DAC value to the DAC
  time.sleep(0.5)  # Delay for half a second
  adc_value = adc.read()  # Read the ADC value (0-4095 for 12-bit width)
  adc_voltage = adc_value / 4095 * 3.3  # Convert ADC value to voltage (based on 3.3V range)
  print("DAC voltage:", dac_voltage, "ADC voltage:", round(adc_voltage, 2))
```
The output is:
```>>> %Run -c $EDITOR_CONTENT

MPY: soft reboot
DAC voltage: 0.25 ADC voltage: 0.17
DAC voltage: 0.5 ADC voltage: 0.43
DAC voltage: 0.75 ADC voltage: 0.63
DAC voltage: 1.0 ADC voltage: 0.88
DAC voltage: 1.25 ADC voltage: 1.11
DAC voltage: 1.5 ADC voltage: 1.36
DAC voltage: 1.75 ADC voltage: 1.6
DAC voltage: 2.0 ADC voltage: 1.83
DAC voltage: 2.25 ADC voltage: 2.06
DAC voltage: 2.5 ADC voltage: 2.31
DAC voltage: 2.75 ADC voltage: 2.56
DAC voltage: 3.0 ADC voltage: 2.87
DAC voltage: 3.25 ADC voltage: 3.29
>>> 
```
If I try to run the program again, I get the error listed above. I believe the error comes from trying to allocate pin 25 again to the DAC. If I edit the program and try to run it again, I get the same error. I have spoken to other developers, and no one seems to be pulling the power on their ESP32 to re-run a program. I tried flashing an older micropython, and that did not work. I had planned to put the DAC specific code for my project in a class, and then call the class as needed. But the program fails because of the allocation of pin 25 to the DAC object. I have tried using the Arduino micropython IDE and Thonny IDE, and no combination of reset commands solves the problem.

I hope someone has a suggestion on how to fix this issue.

**Comment by davefes** [comment]
Before the `break` try `dac.deinit()`

**Reply by davefes** [reply]
If only doing a `soft_reset()` hardware can be left initialised.

I had a lot of "fun" getting cameras to behave:
https://github.com/davefes/ESP32-CAM-Micropython/blob/main/ESP32_CAM_poweron.pdf

Edit:  I meant machine.reset() on ESP32-CAM seemed to leave the camera in an "unknown" state.

**Reply by pmi123** [reply]
The program throws an exception:
```
>>> %Run -c $EDITOR_CONTENT

MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 35, in <module>
AttributeError: 'DAC' object has no attribute 'deinit'
>>> 
```
I tried both dac.deinit() and DAC.deinit(). It also does not have the init() function.

Perhaps I should ditch the onboard DAC and add an external one and use the SPI bus.

**Reply by davefes** [reply]
Oops ... https://docs.micropython.org/en/latest/library/pyb.DAC.html
Maybe, PYboards and any STM32 with an internal DAC only.

**Comment by peterhinch** [comment]
An option might be to call `machine.reset()` at the end of program execution.

**Comment by shariltumin** [comment]
The esp errors were:
```
E (306886) dac_common: dac channel 0 has been registered by dac oneshot
E (306886) dac_oneshot: dac_oneshot_new_channel(45): register dac channel 0 failed
```

You must perform a hard reset after a DAC is registered. A soft reset is not enough. Try the following:

```py

from machine import Pin, ADC, DAC, reset

try:
  dac = DAC(Pin(25))
except OSError:
  print("Unable to register DAC - need hard reset")
  reset()

print(dac)

```

There is a way to delete the channel in esp, ```dac_oneshot_del_channel()```, but this is not implemented in MicroPython (yet?). So you just have to call ```machine.reset()```.

**Reply by davefes** [reply]
I edited my previous response about soft_reset() as my issue was that machine.reset() did not seem to leave the camera in a "good" state.

**Reply by peterhinch** [reply]
I would investigate whether there is a way to issue a hardware reset to the camera.

**Reply by davefes** [reply]
There is a `camera.deinit()` in the distro I am using.  However, that seemed to work for awhile until the camera locked-up and needed a hard reset.  Since using a "poweron reset" circuit I haven't missed a snapshot in months.
